### PR TITLE
Enable batch dense to be used as matrix type for batch solvers

### PR DESCRIPTION
**Pull request by Slaedr**
- Adds missing batch dense kernels.
- Enables testing of solvers with different batch matrix types.
- Enables benchmarking of solvers with batch dense as well.

### CODE REVIEW

File: benchmark/utils/formats.hpp [hunk 1]
```diff
@@ -172,6 +172,7 @@ using hybrid = gko::matrix::Hybrid<etype, itype>;
 using csr = gko::matrix::Csr<etype, itype>;
 using batch_csr = gko::matrix::BatchCsr<etype>;
 using batch_ell = gko::matrix::BatchEll<etype>;
+using batch_dense = gko::matrix::BatchDense<etype>;
```
Comment (Slaedr) [inline benchmark/utils/formats.hpp:175]: I guess this is not actually needed; I'll remove it next time.

File: core/test/utils/batch.hpp [hunk 2]
```diff
@@ -268,10 +278,58 @@ BatchSystem<ValueType> generate_solvable_batch_system(
...
+        mtx->write(mdata2);
+        if (mdata.size() != mdata2.size()) {
+            throw std::runtime_error("sizes not same");
+        }
+        for (size_t ib = 0; ib < mdata.size(); ib++) {
+            if (mdata[ib].nonzeros != mdata2[ib].nonzeros) {
+                throw std::runtime_error("Not same");
+            }
+        }
+    }
```
Comment (Slaedr) [inline core/test/utils/batch.hpp:299]: Remove

File: cuda/solver/batch_bicgstab_kernels.cu [hunk 3]
```diff
@@ -154,7 +154,8 @@ public:
         const size_type nbatch = a.num_batch;
         const int shared_gap = ((a.num_rows - 1) / 8 + 1) * 8;
 
-        // TODO: Add function to BatchCSR to return storage needed per matrix
+        // TODO: Add function to batch matrix types to return storage needed per
```
Comment (Slaedr) [inline cuda/solver/batch_bicgstab_kernels.cu:157]: Fix

**Comment by Slaedr** [comment]
For getting the matrix storage requirements, I added small functions to the uniform batch structs. It's a bit delicate in terms of performance because these structs are passed to the device kernels, but because it's an inline function it should not matter. I could also add a variable, and that would also probably be fine, but it would at least increase the constant memory usage for holding the arguments. I'm guessing the device compiler should be able to throw away unused inline functions. I also checked the performance on summit for gri30 + csr + bicgstab, and there's no regression.

File: core/solver/batch_bicgstab.cpp [hunk 4]
```diff
@@ -146,13 +145,10 @@ void BatchBicgstab<ValueType>::apply_impl(const BatchLinOp* b,
     if (acsr) {
```
Comment (pratikvn) [inline core/solver/batch_bicgstab.cpp:145]: Maybe this if can also be removed now ?

File: common/cuda_hip/matrix/batch_dense_kernels.hpp.inc [hunk 5]
```diff
@@ -35,6 +35,109 @@ template <typename ValueType>
 using BatchEntry = gko::batch_dense::BatchEntry<ValueType>;
 
 
+template <typename ValueType>
+__device__ __forceinline__ void single_spmv_kernel(
```
Comment (pratikvn) [inline common/cuda_hip/matrix/batch_dense_kernels.hpp.inc:39]: Maybe rename to `singke_apply_kernel` ? It is not sparse in this case :)

**Review by pratikvn — APPROVED** [review]
Really good work! There is some commented code, but otherwise looks good to me.

File: common/cuda_hip/matrix/batch_dense_kernels.hpp.inc (hunk 5 above)
Comment (Slaedr) [reply common/cuda_hip/matrix/batch_dense_kernels.hpp.inc:39]: Yeah, I noticed that, but the name needs to be the same across all matrix formats for the batch solvers to work 😅 Maybe we change it to `single_apply_kernel` everywhere, though for free functions that's not as descriptive. Maybe `single_matvec_kernel`?

Comment (pratikvn) [reply common/cuda_hip/matrix/batch_dense_kernels.hpp.inc:39]: Ah, yeah. True. Maybe we can rename it to `single_apply_kernel` everywhere.

**Pull request merged**